index.js

```diff
@@ -68,7 +69,7 @@ function proxy (uri) {
   setupUrl(this);
 
   // attempt to get a proxying `http.Agent` instance
-  var agent = proxyAgent(uri);
+  var agent = proxyAgent;
```

Comment:
I suspect this is the wrong change to make here—disregarding the `uri` passed in to `proxy()`, but I am still unfamiliar with how proxy-agent works.

Reply:
After looking over v6 of proxy-agent, it appears that ProxyAgent is now focused on returning an agent _solely_ based on the conventional environment variables and does not provide anymore an API for returning an agent configured by passing in a URL string or connection object through code.

Is that correct?

If so, I reckon superagent-proxy's `proxy(url)` method needs its own implementation of an agent factory to instantiate one based on the proxy URL passed in.

Reply:
That is correct. The reason that an explicit URL was removed was because the intent of `ProxyAgent` is that it'll be used in CLI apps, and thus the end user should be in control over what proxy to connect to. Thus, specifying a URL within the code no longer made sense to me.

I suggest that `superagent-proxy` should also drop the parameter in `.proxy()`, and simply apply the `ProxyAgent` instance. Typing this out, I question whether this module even needs to exist anymore, since it's essentially would just boil down to:

```js
const agent = new ProxyAgent();

superagent.agent(agent).…

Reply:
> I suggest that superagent-proxy should also drop the parameter in `.proxy()`

This removal would hurt a little for our use of superagent-proxy. We use `.proxy(a_proxy_url)` to configure an agent when our library's users have opted to set a proxy in code rather than with environment variables.† I am totally open to learning how to do that another way, in service to removing obsolete code.

---
† I _know_, right? I prefer env vars, too, but sometimes devs are gonna insist on deving.

Reply:
Hi folks, the `ProxyAgent` ctor has an argument of type `ProxyAgentOptions`, which includes [getProxyForUrl](https://github.com/TooTallNate/proxy-agents/blob/d5cdaa1b774c699c75b543eb4b112290d261e321/packages/proxy-agent/src/index.ts#L99) property. This is how we migrated to the new version of `proxy-agent` an appcenter-cli: https://github.com/microsoft/appcenter-cli/pull/2387/files#diff-dcbeaf7180359f73aa2d9027ca6bcbcac4c9f02ab01152a0f663ea5725aaeff1R17

Reply:
What do you think about this?

```suggestion
  var agent = uri ? new ProxyAgent({ getProxyForUrl: () => uri }) : proxyAgent;
```

So that superagent can get an agent configured with environment variables:

```js
superagent.post(targelUrl).proxy()
```

… or with a proxy connection given as a parameter:

```js
superagent.post(targetUrl).proxy(proxyUrl)
```

Reply:
LGTM.
@robbkidd do you think this PR can be merged soon? Asking because we are using this lib in [code-push](https://github.com/microsoft/code-push) repo.

Reply:
Also, if you are applying the changes, I would correct it a bit:
```
var agent = uri ? new ProxyAgent({ getProxyForUrl: () => uri }) : new ProxyAgent();
```

And remove `const proxyAgent = new ProxyAgent();` in the beginning, to avoid unnecessary instantiation.

Reply:
> @robbkidd do you think this PR can be merged soon?

I'm new here, not a maintainer, so my thoughts are not very influential on merging.

> ```
> var agent = uri ? new ProxyAgent({ getProxyForUrl: () => uri }) : new ProxyAgent();
> ```
> And remove `const proxyAgent = new ProxyAgent();` in the beginning, to avoid unnecessary instantiation.

I believe that would replace a single, maybe-unnecessary instantiation of ProxyAgent with 𝒩 instantiations of a ProxyAgent all with the same configuration where 𝒩 is the number of times a proxied superagent request is made.

... which I realize now is also what would happen when calling `.proxy(uri)` with the same `uri` over and over. 🤔

Reply:
> ... my thoughts are not very influential on merging.

At the moment, as a user of this library, I'm inclined towards dropping the dependency on superagent-proxy in the code I maintain and to take a dependency on proxy-agent directly. We can instantiate it appropriately based on the configuration given to our code. Pretty much as @TooTallNate [described as a possibility earlier](https://github.com/TooTallNate/superagent-proxy/pull/50#discussion_r1282775270).